ledger/ledger-api-test-tool/src/main/scala/com/daml/ledger/api/testtool/tests/ParticipantPruningIT.scala

```diff
@@ -0,0 +1,601 @@
+            o.getAbsolute == offsetOfFirstSurvivingCheckpoint.getAbsolute),
+          s"first checkpoint offset ${firstCheckpointsAfterPrune.offset} after pruning does not match expected offset ${offsetOfFirstSurvivingCheckpoint}"
+        )
+        assertGrpcError(
+          cannotReadAnymore,
+          Status.Code.NOT_FOUND,
+          Some(Pattern.compile(
+            s"Command completions? request from [0-9a-fA-F]* to [0-9a-fA-F]* overlaps with pruned offset ${offsetToPruneUpTo.getAbsolute}"))
+        )
+
+      }
+  })
+
+  test("PRPruneACS", "Prune succeeds by not affecting active contracts", allocate(SingleParty))(
+    implicit ec => {
+      case Participants(Participant(participant, submitter)) =>
+        for {
+          endOffsetAtTestStart <- participant.currentEnd()
+
+          _ <- populateLedger(participant, submitter, batchesToPopulate)
+
+          createdBefore <- participant.activeContracts(submitter)
+
+          transactionsLookup <- participant.transactionTrees(
+            participant
+              .getTransactionsRequest(parties = Seq(submitter), begin = endOffsetAtTestStart))
+
+          offsetToPruneUpTo = offsetAt[TransactionTree](transactionsLookup, itemsToPrune, _.offset)
+
+          _ <- participant.prune(offsetToPruneUpTo)
+
+          createdAfter <- participant.activeContracts(submitter)
+        } yield {
+          assert(
+            createdBefore == createdAfter,
+            "Pruning should not alter the set of active contracts")
+        }
+    })
+
+  test(
+    "PRPruneTreeByTxId",
+    "Prune succeeds as observed by individual transaction tree lookups",
+    allocate(SingleParty))(implicit ec => {
+    case Participants(Participant(participant, submitter)) =>
+      for {
+        endOffsetAtTestStart <- participant.currentEnd()
+
+        size <- populateLedger(participant, submitter, batchesToPopulate)
+
+        transactionsPerBatch = size / batchesToPopulate
+
+        transactionsLookup <- participant.transactionTrees(
+          participant
+            .getTransactionsRequest(parties = Seq(submitter), begin = endOffsetAtTestStart))
+
+        offsetToPruneUpTo = offsetAt[TransactionTree](transactionsLookup, itemsToPrune, _.offset)
+
+        _ <- participant.prune(offsetToPruneUpTo)
+
+        prunedTransactionIds = Range(itemsToPrune - transactionsPerBatch, itemsToPrune).toVector
+          .map(transactionsLookup(_).transactionId)
+        unprunedTransactionIds = Range(itemsToPrune, itemsToPrune + transactionsPerBatch)
+          .map(transactionsLookup(_).transactionId)
+          .toVector
+
+        prunedTransactionTrees <- Future.sequence(
+          prunedTransactionIds.map(participant.transactionTreeById(_, submitter).failed))
+
+        unprunedTransactionTrees <- Future.sequence(
+          unprunedTransactionIds.map(participant.transactionTreeById(_, submitter)))
+      } yield {
+        assert(
+          prunedTransactionTrees.size == transactionsPerBatch,
```

Comment:
Can this ever fail? `prunedTransactionIds` is constructed from `Range(itemsToPrune - transactionsPerBatch, itemsToPrune)`, which has always size `transactionsPerBatch`.

The real check seems to be that the corresponding `Future.sequence(ids.map(participant.transactionTreeById))` succeeds.

Reply:
You are right that the two size-based asserts are tautologies and as such may be confusing. My intent was to express that the last `n` pruned transactions are now invisible and the subsequent `n` unpruned transactions are still visible. I can get rid of these asserts.